### PR TITLE
Update Developer Guide

### PR DESCRIPTION
**Pull request by johnwhitton**

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -1,2 +1,294 @@
 # Horizon Bridge Documentation and API reference
 
+# Overview
+
+The current state of the project is that we are testing using a Harmony Local Node and brdiging to Ropsten. For the client we will use the CLI and moving forward will use the new UI
+* DAG genertion : Takes several hours to run Ganesha has a machine to do this and shares the latest DAG info from Ropsten using [google drive](https://drive.google.com/file/d/1FqLCO5oc1xDYNMuub7xAqnb6kfohdf-U/view?usp=sharing). The epoch logic is the block Number divided by 30,000 so current Ropsten EPOCH is block 12280236 / 30000 = 409 which is the DAG info shared above. **Moving forward we need to update DAG information for every new EPOCH**
```
Comment (polymorpher) [inline docs/README.md:6]: Each epoch should have 32768 blocks https://docs.harmony.one/home/network/validators/definitions/epoch-transition

File: docs/README.md [hunk 2]
```diff
@@ -1,2 +1,294 @@
...
+* CLI Relayer : Relays the blocks, this is initially written in javascript as a Proof of concept and may be implemented in other languages moving forward. **Once the Relayer has begun we need to continually relay each block**
```
Comment (polymorpher) [inline docs/README.md:7]: Might be worthwhile discussing the differences in frequency and behaviors of how block headers are relayed between Ethereum and Harmony

File: docs/README.md [hunk 3]
```diff
@@ -1,2 +1,294 @@
...
+* Client: The client is used to Process transactions this is done by locking the Token using the TokenLocker.sol contract (e.g. TokenLockerOnEth.Sol AND TokenLockerOnHarmony.sol)
```
Comment (polymorpher) [inline docs/README.md:8]: Is this referring to Ethereum and Harmony light clients? I think they are for adding and verifying block headers from the other side. Token lockers are for managing tokens and verifying transactions and releasing the tokens after a bridge transaction is verified.

File: docs/README.md [hunk 4]
```diff
@@ -1,2 +1,294 @@
...
+Currently only ERC20 are supported. Moving forward ERC721 and ERC1155 as well as operations on smart contracts will also be supported. For now all client transactions will be done using the CLI. Moving forward  the current bridge (bridge.harmony.one) will be migrated to https://bridge-validator-1.web.app/ and jenya also built a fresh frontend for upcoming trustless bridge: https://github.com/harmony-one/horizon-trustless-frontend
+but most likely, we will stick to the first one.
+* Harmony Node: we use a [harmony local node](https://github.com/harmony-one/harmony#dev-docker-image) this can be run via docker. **There is a pull request which needs to be pushed to Harmony Testnet to enable the bridging functionality. [we need this PR to be pushed to testnet (should be done by May 27th, 2022)](https://github.com/harmony-one/harmony/pull/3872)**
```
Comment (polymorpher) [inline docs/README.md:11]: Why is this PR needed? Is it not sufficient to have the block header submitters (relays) to compute the MMR?

File: docs/README.md [hunk 5]
```diff
@@ -1,2 +1,294 @@
...
+* For fully trustless bridge we need the bls signature verification precompile to be available on ethereum [eip-2537](https://eips.ethereum.org/EIPS/eip-2537), however this won't be a blocker, as we can initially do permissioned relayers, later adopt optimistic approach, etc. there are many fallback plans for this.
```
Comment (polymorpher) [inline docs/README.md:12]: Can this be clarified in more detail? Where is BLS signature verification needed, and why isn't there any alternative? What are the optimistic approach and fallback options?

File: docs/README.md [hunk 6]
```diff
@@ -1,2 +1,294 @@
...
+* CLI Relayer : Relays the blocks, this is initially written in javascript as a Proof of concept and may be implemented in other languages moving forward. **Once the Relayer has begun we need to continually relay each block**
+* Client: The client is used to Process transactions this is done by locking the Token using the TokenLocker.sol contract (e.g. TokenLockerOnEth.Sol AND TokenLockerOnHarmony.sol)
+Currently only ERC20 are supported. Moving forward ERC721 and ERC1155 as well as operations on smart contracts will also be supported. For now all client transactions will be done using the CLI. Moving forward  the current bridge (bridge.harmony.one) will be migrated to https://bridge-validator-1.web.app/ and jenya also built a fresh frontend for upcoming trustless bridge: https://github.com/harmony-one/horizon-trustless-frontend
```
Comment (polymorpher) [inline docs/README.md:9]: What are the paths to support ERC721 and ERC1155? It would be nice to have some clarifications or pointers to technical components required and TODO lists

File: docs/README.md (hunk 1 above)
Comment (polymorpher) [inline docs/README.md:6]: What are the DAGs for and how is it generated?

File: docs/README.md [hunk 7]
```diff
@@ -1,2 +1,294 @@
...
+| --- | ------ | ----------------------------- | ----- |
+| 1   | *PASS  | Infrastructure Setup          | Working with Local Net until [PR 3872](https://github.com/harmony-one/harmony/pull/3872) is pushed to Testnet |
+| 2   | TBD    | Ropsten Smart Contract Deploy | | 
+| 3   | TBD    | Harmony Smart Contract Deploy | |
+| 4   | TBD    | Relayer Running               | |
+| 5   | TBD    | End To End Testing            | |
+
+**Migration Strategy**
+* Smart Contract use Hardhat with Typescript and ethers(instead of web3)
+  * Replace all web3 with ethers
```
Comment (polymorpher) [inline docs/README.md:30]: Are other developers comfortable with this?

File: docs/README.md [hunk 8]
```diff
@@ -1,2 +1,294 @@
...
+  * replace all js files with typescript
```
Comment (polymorpher) [inline docs/README.md:31]: Similarly for this - let's make sure it doesn't create too much extra work since we need to launch this fast. Can this be postponed for later?

File: docs/README.md [hunk 9]
```diff
@@ -1,2 +1,294 @@
...
+* docs: new folder for documentation
+* docs/assets => migrated from assets
+* docs/solidity: contains generated solidity documentation
+* deploy: new folder for deployment scripts (using hardhat-deploy and logic from scripts)
+* src: new folder for typescript source files 
+* src/lib: (migrated from scripts)
+* src/cli: (migrated from cli)
+* src/(elc, eprover, eth2hmy-relay): migrated from tools(elc, eprover, eth2hmy-relay)
+
+**RollOut Strategy**
```
Comment (polymorpher) [inline docs/README.md:43]: It would be very helpful to have more unit tests, and comments explaining what each component (and unit test) is doing. Right now it is not apparent that each unit component is functioning as intended, what the corner cases are, what the potential attack vectors are, and how they are defended (through the bridging mechanism). It would be nice to have the unit tests illustrating those

File: docs/README.md (hunk 6 above)
Comment (polymorpher) [inline docs/README.md:9]: It seems the frontend implementation contains only a basic user interface, and does not have the majority of the necessary frontend components (APIs, services, state and session management, error handling, UI usage of APIs, and many others). I think the documentation we put here should reflect some of the details and the state of the frontend, otherwise it may give developers, users, and partners an inaccurate impression of the state of completion

File: docs/README.md [hunk 10]
```diff
@@ -1,2 +1,294 @@
...
+* Complete End To End Testing (using CLI)
+* Write Smart Contract Tests (will use hardhat)
+* Update Smart Contract Documentation and README.md
+* Integrate with Front END UI
```
Comment (polymorpher) [inline docs/README.md:47]: As commented above - it seems some substantial frontend implementation work needs to be done before integration could take place

File: docs/README.md [hunk 11]
```diff
@@ -1,2 +1,294 @@
...
+* Onboard Production Validators (15 Validators to relay blocks - incentives tbd)
```
Comment (polymorpher) [inline docs/README.md:48]: Non-validators can also relay blocks as long as they have access to a valid RPC node. Have we decided on the criteria of inclusion for the initial permissioned set of relayers? Many of them would be submitting redundant block headers and epochs - what are the top 3 things we want to achieve through this initial set of relayers, and what is the roadmap for transitioning into a permissionless set?

File: package.json [hunk 12]
```diff
@@ -4,7 +4,8 @@
   "description": "Harmony soccerplayers.app",
   "main": "truffle.js",
   "scripts": {
-    "libInit": "tar -zxf ethereumjs-ethash-dist.tar.gz -C node_modules/@ethereumjs/ethash"
+    "libInit": "tar -zxf ethereumjs-ethash-dist.tar.gz -C node_modules/@ethereumjs/ethash",
```
Comment (polymorpher) [inline package.json:7]: What is this? Where does the .tar.gz package come from and how do relayers and developers verify they are trustworthy?

File: hardhat.config.js [hunk 13]
```diff
@@ -8,6 +8,7 @@ require('@openzeppelin/hardhat-upgrades');
  */
 
 const HARMONY_PRIVATE_KEY = process.env.PRIVATE_KEY;
+const PROJECT_ID = process.env.INFURA_PROJECT_ID
```
Comment (polymorpher) [inline hardhat.config.js:11]: Infura sometimes goes down for Ethereum testnets - I suggest we allow freeform provider URLs

File: docs/README.md [hunk 14]
```diff
@@ -1,2 +1,294 @@
...
+- [ ] Review item which is needed for Permissionless Rollout of Validators (waiting on something on Ethereum?)
+- [ ] Check test updates from Boris
+
+## Environment Data
+
+Here is an overview of Environment data setup for testing after following the steps below
+
+* Deployer Address : 0x8875fc2A47E35ACD1784bB5f58f563dFE86A8451
+* Infura Project: 32cb9c57bfe447a99ea34e30195b2d10
+* KOVAN ERC20 Contract: [0xc90a6555CaD53a9D85a80052Fe2926E21608CF41](https://kovan.etherscan.io/address/0xc90a6555cad53a9d85a80052fe2926e21608cf41)
```
Comment (polymorpher) [inline docs/README.md:63]: What is this contract for? Any testing facilities on the contract?

File: docs/README.md [hunk 15]
```diff
@@ -1,2 +1,294 @@
...
+
+```
+On localnet, by default, 0xA5241513DA9F4463F1d4874b548dFBAC29D91f34 has funds, as defined in core/genesis.go. The private key for this address is 1f84c95ac16e6a50f08d44c7bde7aff8742212fda6e4321fde48bf83bef266dc
+```
+
+For testing purposes we used account `0x8875fc2A47E35ACD1784bB5f58f563dFE86A8451` and funded it with 1000 ONE on localnet and 1 ETH on Ropsten, but you can use any acccount you like as long as you know the private key or mnemonic.
+
+## Creating the DAG Merkle Tree
+
+DAG genertion takes several hours to run Ganesha has a machine to do this and shares the latest DAG info from Ropsten using [google drive](https://drive.google.com/file/d/1FqLCO5oc1xDYNMuub7xAqnb6kfohdf-U/view?usp=sharing). The epoch logic is the block Number divided by 30,000 so current Ropsten EPOCH is block 12280236 / 30000 = 409 which is the DAG info shared above.
```
Comment (polymorpher) [inline docs/README.md:121]: The code looks like something that can be parallelized and made efficient. What exactly is DAG and what is the purpose of generating all of them? Why is it aligned to epoch numbers? I think these questions should be answered before discussing how DAG is generated and the relevant commands

File: docs/README.md [hunk 16]
```diff
@@ -1,2 +1,294 @@
...
+2. Do you need to generate a new DAG information on each EPOC change?
+3. Where is this currently hosted?
+4. Where do we want to deploy this long term?
+
+## Deploying Smart Contracts
+
+Following is an overview of the contracts used in the bridge and which chain they should be deployed on. *Note: her we only focus on the contracts deployed the imported contracts are not covered*
+
+**Ethereum (Kovan)**
+* HarmonyLightClient.sol : stores all blockheaders which are used for verification.
```
Comment (polymorpher) [inline docs/README.md:161]: Harmony block headers - I think it would be informative to document the differences compared to Ethereum block headers, and how the light clients act differently (frequency and nature of updates, cost considerations, etc.)

File: docs/README.md [hunk 17]
```diff
@@ -1,2 +1,294 @@
...
+* TokenLockerOnEthereum.sol : Tracks tokens locked which are then minted by the bridge
```
Comment (polymorpher) [inline docs/README.md:162]: Responsible for proof submission (i.e. execution of releasing tokens) as well

File: docs/README.md [hunk 18]
```diff
@@ -1,2 +1,294 @@
...
+1. `node index.js dagProve generate` which calculate merkle root for epochs from [start, start+n)
+```
+node index.js dagProve generate 377
+```
+2. `node index.js dagProve blockProof` which accepts block number to calculate all necessary information in order to prove the block
+```
+node index.js dagProve blockProof --block 11266784 --url https://ropsten.infura.io/v3/<project-id>
+```
+
+**Questions**
```
Comment (polymorpher) [inline docs/README.md:149]: Move elsewhere?

File: docs/README.md [hunk 19]
```diff
@@ -1,2 +1,294 @@
...
+7. `node index.js Bridge crossBack` cross transfer HRC20 from harmony back to ethereum.
+
+There are also two scripts written to facilitate testing. 
+1. [test.js](https://github.com/johnwhitton/horizon/blob/main/scripts/test.js)
+2. [newtests.js](https://github.com/johnwhitton/horizon/blob/main/scripts/newtest.js)
+
+Also [end2end.js](https://github.com/johnwhitton/horizon/blob/main/scripts/end2end.js) which does not have any code in it.
+
+**Original Documentation**
+Original documentation for the CLI contains [Bridge CLI](https://github.com/johnwhitton/horizon/tree/main/cli#bridge-cli), [Ethereum Receipt Prove CLI](https://github.com/johnwhitton/horizon/tree/main/cli#ethereum-receipt-prove-cli) and [Ethereum Receipt Verifier CLI](https://github.com/johnwhitton/horizon/tree/main/cli#ethereum-receipt-verifier-cli).
```
Comment (polymorpher) [inline docs/README.md:224]: Can all links point to files in the original repository and made permalink (so they don't change across versions?)

File: docs/README.md [hunk 20]
```diff
@@ -1,2 +1,294 @@
...
+
+## End to End Testing
+End to End testing currently consists of mapping ERC20 tokens from Harmony to Ethereum and transferring them accross the chains. This is done via the [ClI](https://github.com/johnwhitton/horizon/tree/main/cli) using the following commands
+
+5. `node index.js Bridge map` map ERC20 from ethereum to harmony.
+6. `node index.js Bridge crossTo` cross transfer ERC20 from ethereum to harmony.
+7. `node index.js Bridge crossBack` cross transfer HRC20 from harmony back to ethereum.
+
+There are also two scripts written to facilitate testing. 
+1. [test.js](https://github.com/johnwhitton/horizon/blob/main/scripts/test.js)
```
Comment (polymorpher) [inline docs/README.md:218]: This seems to be just a few unit tests with some utility functions

File: docs/README.md [hunk 21]
```diff
@@ -1,2 +1,294 @@
...
+2. [newtests.js](https://github.com/johnwhitton/horizon/blob/main/scripts/newtest.js)
```
Comment (polymorpher) [inline docs/README.md:219]: This seems to trigger an actual "bridging" but I am not sure if it is supposed to work. It covers some deployment and verification stuff, but doesn't seem to cover end-to-end

File: docs/README.md [hunk 22]
```diff
@@ -1,2 +1,294 @@
...
+
+Also [end2end.js](https://github.com/johnwhitton/horizon/blob/main/scripts/end2end.js) which does not have any code in it.
+
+**Original Documentation**
+Original documentation for the CLI contains [Bridge CLI](https://github.com/johnwhitton/horizon/tree/main/cli#bridge-cli), [Ethereum Receipt Prove CLI](https://github.com/johnwhitton/horizon/tree/main/cli#ethereum-receipt-prove-cli) and [Ethereum Receipt Verifier CLI](https://github.com/johnwhitton/horizon/tree/main/cli#ethereum-receipt-verifier-cli).
+
+The assumption is that you need to just run the map, crossTo and crossBack functions.
+
+**Questions**
```
Comment (polymorpher) [inline docs/README.md:228]: Move elsewhere? As to the questions, I suppose Bridge subcommands (the three you mentioned) can be used to do human-driven testing once deployment is complete. But they don't seem to be good for any automated testing. I don't think there is any end-to-end testing elsewhere

File: docs/README.md [hunk 23]
```diff
@@ -1,2 +1,294 @@
...
+
+## Smart Contract Testing
+Smart contract testing has a [test folder](https://github.com/johnwhitton/horizon/tree/main/test) which contains [bridg.hmy.js](https://github.com/johnwhitton/horizon/blob/main/test/bridge.hmy.js) however at the time of writing this test was not working.
+
+Tests are set up in the tests folder and are run using hardhat.
+
+`yarn test` is set up as a script to run `npx hardhat test`
+
+**Questions**
+1. Is there any other branches where tests have been written?
```
Comment (polymorpher) [inline docs/README.md:243]: Some PRs seem to be introducing useful tests, for example: https://github.com/harmony-one/horizon/pull/31/files

File: docs/README.md [hunk 24]
```diff
@@ -1,2 +1,294 @@
...
+2. Are there any dependencies on tools or utils for these tests? 
+
+# Component Overview
```
Comment (polymorpher) [inline docs/README.md:246]: This would be quite useful - let's make this a standalone doc

File: docs/README.md [hunk 25]
```diff
@@ -1,2 +1,294 @@
...
+| 4.4     | upgrade                     |
+| 4.5     | configure.js                |
+| 4.6     | test.js                     |
+| 4.7     | newtest.js                  |
+| 4.8     | end2end.js                  |
+| **5**   | **test**                    |
+| 5.1     | bridge.hmy.js               |
+
+
+**Migration Strategy**
```
Comment (polymorpher) [inline docs/README.md:281]: repeating some stuff above? see earlier "Migration Strategy"

**Comment by polymorpher** [comment]
Should we break this down to smaller pieces and target a non-main branch first?
